x/application/keeper/msg_server_undelegate_from_gateway_test.go

```diff
@@ -272,3 +281,225 @@ func TestMsgServer_UndelegateFromGateway_SuccessfullyUndelegateFromUnstakedGatew
+
+	// Get the updated application state after pruning logic is run.
+	app, isAppFound = k.GetApplication(sdkCtx, app.Address)
+	require.True(t, isAppFound)
+	require.NotNil(t, app)
+
+	// Verify that the application is still delegated to the gateway
+	require.Contains(t, app.DelegateeGatewayAddresses, gatewayAddrToRedelegate)
+
+	// Verify that the the pending undelegation map no longer contains the
+	// sessionEndHeight key.
+	require.Empty(t, app.PendingUndelegations[sessionEndHeight])
+
+	// Verify that the reconstructed delegatee gateway list includes the reundelegated gateway
+	gatewayAddressesAfterPruning := rings.GetRingAddressesAtBlock(&app, sdkCtx.BlockHeight())
+	require.Contains(t, gatewayAddressesAfterPruning, gatewayAddrToRedelegate)
+}
+
+// delegateThenUndelegate factors delegating an application to two gateways,
+// then undelegates the application from one of the gateways to be used in
+// multiple tests.
+func delegateThenUndelegate(
+	ctx context.Context,
+	t *testing.T,
+	srv types.MsgServer,
+	k keeper.Keeper,
+	undelegationHeight int64,
+) (sdkCtx sdk.Context, app types.Application, gatewayAddrs []string) {
+	// Generate an application address and stake the application.
+	appAddr := sample.AccAddress()
+	stakeMsg := &types.MsgStakeApplication{
+		Address: appAddr,
+		Stake:   &sdk.Coin{Denom: "upokt", Amount: math.NewInt(100)},
+		Services: []*sharedtypes.ApplicationServiceConfig{
+			{
+				Service: &sharedtypes.Service{Id: "svc1"},
+			},
+		},
+	}
+	_, err := srv.StakeApplication(ctx, stakeMsg)
+	require.NoError(t, err)
+
+	// Generate gateway addresses, mock the gateways being staked then delegate the
+	// application to the gateways.
+	for i := 0; i < 2; i++ {
+		gatewayAddr := sample.AccAddress()
+		gatewayAddrs = append(gatewayAddrs, gatewayAddr)
+
+		keepertest.AddGatewayToStakedGatewayMap(t, gatewayAddr)
+
+		delegateMsg := &types.MsgDelegateToGateway{
+			AppAddress:     appAddr,
+			GatewayAddress: gatewayAddr,
+		}
+		_, err = srv.DelegateToGateway(ctx, delegateMsg)
+		require.NoError(t, err)
+	}
+
+	gatewayAddrToUndelegate := gatewayAddrs[0]
+
+	// Create a context with a block height of 2.
+	sdkCtx = sdk.UnwrapSDKContext(ctx).WithBlockHeight(undelegationHeight)
+
+	// Undelegate from the first gateway
+	undelegateMsg := &types.MsgUndelegateFromGateway{
+		AppAddress:     appAddr,
+		GatewayAddress: gatewayAddrToUndelegate,
+	}
+	_, err = srv.UndelegateFromGateway(sdkCtx, undelegateMsg)
+	require.NoError(t, err)
+
+	foundApp, isAppFound := k.GetApplication(sdkCtx, appAddr)
+
+	// Verify that the application exists.
+	require.True(t, isAppFound)
+	require.NotNil(t, foundApp)
+
+	return sdkCtx, foundApp, gatewayAddrs
+}
+
+func getPruningBlockHeight(currentHeight int64) (pruningHeihgt int64) {
+	nextSessionStartHeight := int64(sessionkeeper.GetSessionEndBlockHeight(currentHeight) + 1)
+
+	numBlocksUndelegationRetention := sessionkeeper.GetSessionGracePeriodBlockCount() +
```

Comment:
Isn't there a function / helper for this in the actual source code?

Reply:
Added.